In [1]:
from pandas import read_csv
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import numpy as np

In [42]:
df = read_csv('data\\housing_data_na.csv')
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2

In [7]:
df.shape
df.isna().sum()

CRIM       20
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

In [43]:
df.dropna(inplace=True)

In [9]:
# tuong quan giua cac bien
df.corr()

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.191178  0.401863 -0.054355  0.417130 -0.219150  0.354342   
ZN      -0.191178  1.000000 -0.531871 -0.037229 -0.513704  0.320800 -0.563801   
INDUS    0.401863 -0.531871  1.000000  0.059859  0.764866 -0.390234  0.638431   
CHAS    -0.054355 -0.037229  0.059859  1.000000  0.075097  0.104885  0.078831   
NOX      0.417130 -0.513704  0.764866  0.075097  1.000000 -0.302188  0.731548   
RM      -0.219150  0.320800 -0.390234  0.104885 -0.302188  1.000000 -0.247337   
AGE      0.354342 -0.563801  0.638431  0.078831  0.731548 -0.247337  1.000000   
DIS     -0.374166  0.656739 -0.711709 -0.093971 -0.769230  0.205246 -0.744844   
RAD      0.624765 -0.310919  0.604533  0.001468  0.611441 -0.209847  0.458349   
TAX      0.580595 -0.312371  0.731055 -0.032304  0.668023 -0.292048  0.509114   
PTRATIO  0.281110 -0.414046  0.390954 -0.111304  0.188933 -0.355501  0.269226   
B       -0.381411  0.171303 -0.360532  0.051264 -0.380051  0.128069 -0.275303   
LSTAT    0.444943 -0.414193  0.590690 -0.047424  0.582641 -0.614339  0.602891   
MEDV    -0.391363  0.373136 -0.481772  0.181391 -0.427321  0.695360 -0.394656   

              DIS       RAD       TAX   PTRATIO         B     LSTAT      MEDV  
CRIM    -0.374166  0.624765  0.580595  0.281110 -0.381411  0.444943 -0.391363  
ZN       0.656739 -0.310919 -0.312371 -0.414046  0.171303 -0.414193  0.373136  
INDUS   -0.711709  0.604533  0.731055  0.390954 -0.360532  0.590690 -0.481772  
CHAS    -0.093971  0.001468 -0.032304 -0.111304  0.051264 -0.047424  0.181391  
NOX     -0.769230  0.611441  0.668023  0.188933 -0.380051  0.582641 -0.427321  
RM       0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.614339  0.695360  
AGE     -0.744844  0.458349  0.509114  0.269226 -0.275303  0.602891 -0.394656  
DIS      1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.493328  0.249929  
RAD     -0.494588  1.000000  0.910228  0.464741 -0.444413  0.479541 -0.381626  
TAX     -0.534432  0.910228  1.000000  0.460853 -0.441808  0.536110 -0.468536  
PTRATIO -0.232471  0.464741  0.460853  1.000000 -0.177383  0.375966 -0.507787  
B        0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.369889  0.333461  
LSTAT   -0.493328  0.479541  0.536110  0.375966 -0.369889  1.000000 -0.735822  
MEDV     0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.735822  1.000000

In [ ]:
df_X = df.drop(columns=['MEDV'])

In [26]:
X = df.drop(columns=['MEDV']).values
y = df['MEDV'].values
X.shape, y.shape

((394, 13), (394,))

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X,y)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((295, 13), (99, 13), (295,), (99,))

In [28]:
lg = LinearRegression()
lg.fit(X_train, y_train)


LinearRegression()

In [32]:
# He so hoi qui tuyen tinh, he so bi chan
lg.coef_, lg.intercept_

(array([-5.52667406e-02,  4.70814044e-02, -1.14817121e-02,  3.05430019e+00,
        -1.65734236e+01,  4.76275161e+00, -2.42393530e-02, -1.52901849e+00,
         2.44109934e-01, -1.11469354e-02, -9.12858602e-01,  1.29822166e-02,
        -3.74930997e-01]),
 np.float64(28.105799066697852))

In [31]:
lg.score(X_train, y_train), lg.score(X_test, y_test)

(0.7722138900290898, 0.7266795554189628)

In [12]:
## kiem dinh dung stat model
from statsmodels.formula.api import ols

In [14]:
model = ols('MEDV ~ CRIM + ZN + INDUS + CHAS + NOX + RM + AGE + DIS + RAD + TAX + PTRATIO + B + LSTAT', df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.767
Model:                            OLS   Adj. R-squared:                  0.759
Method:                 Least Squares   F-statistic:                     96.29
Date:                Sun, 02 Mar 2025   Prob (F-statistic):          1.75e-111
Time:                        14:50:19   Log-Likelihood:                -1143.4
No. Observations:                 394   AIC:                             2315.
Df Residuals:                     380   BIC:                             2370.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     32.6801      5.681      5.752      0.000      21.509      43.851
CRIM          -0.0976      0.032     -3.007      0.003      -0.161      -0.034
ZN             0.0489      0.014      3.397      0.001       0.021       0.077
INDUS          0.0304      0.066      0.461      0.645      -0.099       0.160
CHAS           2.7694      0.925      2.993      0.003       0.950       4.588
NOX          -17.9690      4.243     -4.235      0.000     -26.311      -9.627
RM             4.2833      0.471      9.100      0.000       3.358       5.209
AGE           -0.0130      0.014     -0.898      0.370      -0.041       0.015
DIS           -1.4585      0.211     -6.912      0.000      -1.873      -1.044
RAD            0.2859      0.069      4.125      0.000       0.150       0.422
TAX           -0.0131      0.004     -3.324      0.001      -0.021      -0.005
PTRATIO       -0.9146      0.141     -6.506      0.000      -1.191      -0.638
B              0.0097      0.003      3.251      0.001       0.004       0.015
LSTAT         -0.4237      0.055     -7.700      0.000      -0.532      -0.315
==============================================================================
Omnibus:                      161.243   Durbin-Watson:                   1.247
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              904.814
Skew:                           1.657   Prob(JB):                    3.33e-197
Kurtosis:                       9.643   Cond. No.                     1.57e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.57e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

# Giả thiết
$$ H_0: X_i = 0$$
$$ H_1: X_i != 0$$
Biến INDUS và AGE có p-value > 0.05 Không thể bác bỏ giả thiết H0 -> tức 2 biến IDUX và AGE ko có ý nghĩa trong mô hình

In [33]:
## Kiem tra da cong tuyen multicollinearity
model = ols('CRIM ~  ZN + INDUS + CHAS + NOX + RM + AGE + DIS + RAD + TAX + PTRATIO + B + LSTAT', df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   CRIM   R-squared:                       0.426
Model:                            OLS   Adj. R-squared:                  0.408
Method:                 Least Squares   F-statistic:                     23.54
Date:                Sun, 02 Mar 2025   Prob (F-statistic):           4.21e-39
Time:                        15:29:11   Log-Likelihood:                -1323.8
No. Observations:                 394   AIC:                             2674.
Df Residuals:                     381   BIC:                             2725.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     12.4250      8.945      1.389      0.166      -5.162      30.012
ZN             0.0360      0.023      1.589      0.113      -0.009       0.081
INDUS         -0.0492      0.104     -0.473      0.636      -0.254       0.155
CHAS          -1.8462      1.457     -1.267      0.206      -4.711       1.019
NOX           -9.1165      6.681     -1.365      0.173     -22.252       4.019
RM            -0.3064      0.743     -0.413      0.680      -1.767       1.154
AGE            0.0032      0.023      0.141      0.888      -0.042       0.048
DIS           -0.7491      0.331     -2.264      0.024      -1.400      -0.099
RAD            0.5890      0.105      5.603      0.000       0.382       0.796
TAX           -0.0031      0.006     -0.504      0.614      -0.015       0.009
PTRATIO       -0.1299      0.222     -0.586      0.558      -0.566       0.306
B             -0.0111      0.005     -2.387      0.017      -0.020      -0.002
LSTAT          0.2543      0.086      2.961      0.003       0.085       0.423
==============================================================================
Omnibus:                      528.195   Durbin-Watson:                   1.648
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            59270.262
Skew:                           6.615   Prob(JB):                         0.00
Kurtosis:                      61.612   Cond. No.                     1.57e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.57e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [34]:
model = ols('NOX ~ CRIM + ZN + INDUS + CHAS + RM + AGE + DIS + RAD + TAX + PTRATIO + B + LSTAT', df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    NOX   R-squared:                       0.778
Model:                            OLS   Adj. R-squared:                  0.771
Method:                 Least Squares   F-statistic:                     111.0
Date:                Sun, 02 Mar 2025   Prob (F-statistic):          2.83e-116
Time:                        15:29:49   Log-Likelihood:                 596.23
No. Observations:                 394   AIC:                            -1166.
Df Residuals:                     381   BIC:                            -1115.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.7380      0.057     12.892      0.000       0.625       0.851
CRIM          -0.0005      0.000     -1.365      0.173      -0.001       0.000
ZN            -0.0002      0.000     -1.043      0.298      -0.001       0.000
INDUS          0.0043      0.001      5.684      0.000       0.003       0.006
CHAS           0.0029      0.011      0.264      0.792      -0.019       0.025
RM            -0.0117      0.006     -2.067      0.039      -0.023      -0.001
AGE            0.0009      0.000      5.502      0.000       0.001       0.001
DIS           -0.0139      0.002     -5.667      0.000      -0.019      -0.009
RAD            0.0034      0.001      4.177      0.000       0.002       0.005
TAX         2.638e-05   4.77e-05      0.552      0.581   -6.75e-05       0.000
PTRATIO       -0.0104      0.002     -6.483      0.000      -0.014      -0.007
B           -6.07e-05   3.57e-05     -1.699      0.090      -0.000    9.54e-06
LSTAT          0.0005      0.001      0.732      0.465      -0.001       0.002
==============================================================================
Omnibus:                       54.924   Durbin-Watson:                   0.511
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               97.321
Skew:                           0.815   Prob(JB):                     7.36e-22
Kurtosis:                       4.809   Cond. No.                     1.18e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.18e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [44]:
## tinh vif: he so phong dai
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [48]:
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 2.852309927713017),
 ('ZN', 14.263621604094793),
 ('INDUS', 1.147676919923875),
 ('CHAS', 78.19587146837218),
 ('NOX', 83.38739667510177),
 ('RM', 22.40573051836839),
 ('AGE', 14.367438543452744),
 ('DIS', 13.986225022429808),
 ('RAD', 58.485821213801586),
 ('TAX', 88.48314836650732),
 ('PTRATIO', 21.04718393865385),
 ('B', 11.340925461230494)]

## ('TAX', 88.48314836650732), --> loai bo bien TAX do vid > 10
## chay lan 2 voi data bo di TAX

In [49]:
df_X.drop(columns=['TAX'], inplace=True)

In [50]:

[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]


[('CRIM', 2.6809703845178126),
 ('ZN', 11.082320504958798),
 ('INDUS', 1.1288316603415711),
 ('CHAS', 77.27672049459568),
 ('NOX', 83.30506442190782),
 ('RM', 22.40105035728029),
 ('AGE', 14.354150509952007),
 ('DIS', 5.469152266527214),
 ('RAD', 86.02531275838574),
 ('PTRATIO', 21.035663798445256),
 ('B', 11.278524392198662)]

('RAD', 86.02531275838574), --> loai tiep RAD

In [52]:
df_X.drop(columns=['RAD'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 2.6497403363393515),
 ('ZN', 10.898046086933322),
 ('INDUS', 1.125069148887821),
 ('CHAS', 73.08844793565146),
 ('NOX', 82.15455900670372),
 ('RM', 22.203620847652896),
 ('AGE', 14.188679413092112),
 ('DIS', 80.97959369651392),
 ('PTRATIO', 19.465140632256663),
 ('B', 11.25992095445878)]

('NOX', 82.15455900670372),

In [53]:
df_X.drop(columns=['NOX'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 2.6496353242777375),
 ('ZN', 8.87938411961582),
 ('INDUS', 1.1243925587979051),
 ('CHAS', 61.643838696061195),
 ('RM', 20.788917612904427),
 ('AGE', 13.937206280671514),
 ('DIS', 80.30162335212944),
 ('PTRATIO', 19.39127635901592),
 ('B', 10.277196711892255)]

In [54]:
df_X.drop(columns=['DIS'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 2.14527369872214),
 ('ZN', 7.760376227345311),
 ('INDUS', 1.1218623535573093),
 ('CHAS', 60.386153813073506),
 ('RM', 17.524782312300434),
 ('AGE', 66.62027590501366),
 ('PTRATIO', 19.10020702627132),
 ('B', 10.156728904063435)]

In [55]:
df_X.drop(columns=['AGE'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 1.8118563989685619),
 ('ZN', 6.834960049110824),
 ('INDUS', 1.1196282086272837),
 ('CHAS', 49.27887951531067),
 ('RM', 64.71854016464373),
 ('PTRATIO', 19.08711835637532),
 ('B', 8.144962588545111)]

In [56]:
df_X.drop(columns=['RM'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 1.587728171928802),
 ('ZN', 6.811950744337582),
 ('INDUS', 1.0981456217701757),
 ('CHAS', 29.094354539277614),
 ('PTRATIO', 17.33493663056349),
 ('B', 7.507229261968338)]

In [57]:
df_X.drop(columns=['CHAS'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 1.5875845279596006),
 ('ZN', 6.724173749295366),
 ('INDUS', 28.92182564872432),
 ('PTRATIO', 17.058827897937178),
 ('B', 7.501715896153426)]

In [58]:
df_X.drop(columns=['INDUS'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 1.4139636901597321),
 ('ZN', 25.657967207511938),
 ('PTRATIO', 16.924431882130072),
 ('B', 6.5027822821498695)]

In [59]:
df_X.drop(columns=['ZN'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('CRIM', 25.643512083958324),
 ('PTRATIO', 16.05961001350894),
 ('B', 6.08333873959481)]

In [60]:
df_X.drop(columns=['CRIM'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('PTRATIO', 14.690572172265504), ('B', 5.463870060540619)]

In [61]:
df_X.drop(columns=['PTRATIO'], inplace=True)
[(df_X.columns[i-1], float(variance_inflation_factor(df_X,i))) for i in range(1, df_X.shape[1])]

[('B', 2.735613366974735)]